File: R-Tree.cpp
#include <iostream>
#include "RTree.h"

void showMenu() {
    std::cout << "R-tree Operations Menu:\n";
    std::cout << "1. Insert MBR\n";
    std::cout << "2. Search MBR\n";
    std::cout << "3. Find Area of MBRs\n";
    std::cout << "4. Display Tree\n"; // New option
    std::cout << "5. Exit\n";        // Adjusted numbering
    std::cout << "Enter your choice: ";
}

int main() {
    RTree rtree(4);
    int choice;
    int x_min, y_min, x_max, y_max;

    while (true) {
        showMenu();
        std::cin >> choice;

        switch (choice) {
        case 1:  // Insert MBR
            std::cout << "Enter MBR coordinates (x_min, y_min, x_max, y_max): ";
            std::cin >> x_min >> y_min >> x_max >> y_max;
            if (x_min > x_max || y_min > y_max) {
                std::cout << "Invalid MBR: x_min must be <= x_max and y_min must be <= y_max\n";
            }
            else {
                rtree.insert(MBR(x_min, y_min, x_max, y_max));
                std::cout << "MBR inserted successfully.\n";
            }
            break;

        case 2:  // Search MBR
            std::cout << "Enter search MBR coordinates (x_min, y_min, x_max, y_max): ";
            std::cin >> x_min >> y_min >> x_max >> y_max;
            if (x_min > x_max || y_min > y_max) {
                std::cout << "Invalid search MBR: x_min must be <= x_max and y_min must be <= y_max\n";
            }
            else {
                rtree.search(MBR(x_min, y_min, x_max, y_max));
            }
            break;

        case 3:  // Find Area of MBRs
            rtree.find_area();
            break;

        case 4:  // Display Tree
            rtree.display_tree();
            break;

        case 5:  // Exit
            std::cout << "Exiting program.\n";
            return 0;

        default:
            std::cout << "Invalid choice. Please try again.\n";
        }
    }

    return 0;
}

File: rtree.py
class MBR:
    def __init__(self, x_min=0, y_min=0, x_max=0, y_max=0):
        self.x_min = x_min
        self.y_min = y_min
        self.x_max = x_max
        self.y_max = y_max
    
    def area(self):
        return (self.x_max - self.x_min) * (self.y_max - self.y_min)
    
    def merge(self, other):
        return MBR(
            min(self.x_min, other.x_min),
            min(self.y_min, other.y_min),
            max(self.x_max, other.x_max),
            max(self.y_max, other.y_max)
        )
    
    def contains(self, other):
        return (self.x_min <= other.x_min and self.y_min <= other.y_min and
                self.x_max >= other.x_max and self.y_max >= other.y_max)
    
    def intersects(self, other):
        return not (self.x_max < other.x_min or self.x_min > other.x_max or
                    self.y_max < other.y_min or self.y_min > other.y_max)

class Node:
    def __init__(self, is_leaf=False):
        self.is_leaf = is_leaf
        self.mbr = MBR()
        self.children = []

class RTree:
    def __init__(self, max_children=4):
        self.max_children = max_children
        self.root = Node(True)
    
    def insert(self, mbr):
        pass  # Implemented in functions.py
    
    def search(self, search_mbr):
        pass  # Implemented in functions.py
    
    def find_area(self):
        pass  # Implemented in functions.py
    
    def display_tree(self):
        pass  # Implemented in functions.py
    
    def insert_recursive(self, node, mbr, parent=None):
        pass  # Implemented in functions.py
    
    def choose_best_node(self, node, mbr):
        pass  # Implemented in functions.py
    
    def split_node(self, node, parent):
        pass  # Implemented in functions.py
    
    def update_mbr(self, node):
        pass  # Implemented in functions.py
    
    def search_recursive(self, node, search_mbr, results):
        pass  # Implemented in functions.py
    
    def find_area_recursive(self, node, level):
        pass  # Implemented in functions.py
    
    def display_tree_recursive(self, node, level=0):
        pass  # Implemented in functions.py